openssl/src/pkcs12.rs

```diff
@@ -164,17 +164,17 @@ impl Pkcs12Builder {
         self,
         password: &str,
         friendly_name: &str,
-        pkey: &PKeyRef<T>,
-        cert: &X509Ref,
+        pkey: Option<&PKeyRef<T>>,
```

Comment:
The parameter type here could be `Into<Option<...>>` so that the existing code continues to work and `None` is a new allowed value to be passed as this parameter.

More elaborate discussion of this technique is in http://web.archive.org/web/20170930063008/http://www.suspectsemantics.com/blog/2016/11/29/the-into-trick/

As for real code using this see: https://gitlab.com/sequoia-pgp/sequoia/-/blob/main/openpgp/src/cert/builder.rs#L443

Reply:
Yeah I didn't mean it will be a silver-bullet, just more backwards-compatibility-proof.

> As far as I can tell, you'd still lose the ability to automatically ref a `PKey<_>` as `&PKeyRef<_>` without defining the trait `From<&pkey::PKey<_>>` for `Option<&PKeyRef<_>>`.

Another option (no pun intended) would be to accept an `Into<Option<AsRef` parameter and impl that for `PKey(Ref)`... but at this point the function signature looks extremely ugly.

> So calling code may have to change, do you still think the change should stand?

I'm just thinking about options and trade-offs. I don't have any strong opinion either way. (Sorry if that's not the answer you were looking for :pray:).

Reply:
> Another option (no pun intended) would be to accept an `Into<Option<AsRef` parameter and impl that for `PKey(Ref)`... but at this point the function signature looks extremely ugly.

I've done this where the function signature becomes (can't seem to stack the traits - if you know how I can, please tell me...):
```rust
pub fn build<T, U, V, W, X>(
    self,
    password: &str,
    friendly_name: &str,
    pkey: U,&pkey
    cert: V,
) -> Result<Pkcs12, ErrorStack>
where
    T: HasPrivate,
    U: Into<Option<W>>,
    V: Into<Option<X>>,
    W: AsRef<PKeyRef<T>>,
    X: AsRef<X509Ref>,
```
That can desugar the following types/examples for `pkey`, (generic ignored):
* `PKey`
* `&PKey`
* `PKeyRef`
* `&PKeyRef`
* `None`
But when passing `None`, you have to define the generic types that it cannot infer:
```rust
let pkcs12 = pkcs12_builder
    .build::<Private, _, _, PKeyRef<_>, _>("mypass", subject_name, None, &cert)
    .unwrap();
```

The equivalent applies to the `cert` parameter.

In my opinion, this seems like a poor trade-off the sake of backwards compatibility :woozy_face:  
To me it becomes a bit less obvious the values that _can_ be passed to it, what do you think?

Reply:
I think the better option would be to add `name`, `pkey` and `cert` setters to the builder type, and then a `build2` method that only takes the truly required parameters. I *think* `pass` is required (?) but a test would confirm.

Reply:
I agree with the setters, that matches with the optional CA stack.

On the assumption you'd like a backwards compatible build function now, I've added a `build2` function in a feature branch in my repo. What do you think? To me it seems like a lot duplication to for the purpose of keeping a backwards compatible `build` function. The setters take a `mut` PKCS12 builder, which wouldn't be passed to the `build` function without changing the signature so they couldn't be used to refactor the backward compatible build function.

Reply:
Changing `self` to `mut self` doesn't change the signature of the method - just what the implementation of the method can do with self. It should be fine to have build delegate to build2.